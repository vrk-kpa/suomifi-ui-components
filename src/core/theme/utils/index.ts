export {
  withSuomifiDefaultProps,
  themeOrTokens,
  TokensOrThemeProps,
} from './defaults';
export { focus } from './focus';
export { boxshadowOutline } from './outline';
export { margin, padding, spacingModifiers } from './spacing';
export { typographyUtils, fontSize } from './typography';
