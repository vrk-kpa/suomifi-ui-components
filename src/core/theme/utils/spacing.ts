import {
  SuomifiTheme,
  SpacingProp,
  defaultSuomifiTheme,
} from '../SuomifiTheme/SuomifiTheme';

export type SpacingWithoutInsetProp =
  | 'xxs'
  | 'xs'
  | 's'
  | 'm'
  | 'l'
  | 'xl'
  | 'xxl'
  | 'xxxl'
  | 'xxxxl'
  | '0';

const spaceVal = (theme: SuomifiTheme) => (val?: SpacingProp) => {
  if (val === '0') return '0';
  return !!val ? theme.spacing[val] : '';
};

export interface PaddingProps {
  /** Padding from theme */
  padding?: SpacingWithoutInsetProp;
  /** Padding-top from theme */
  pt?: SpacingWithoutInsetProp;
  /** Padding-right from theme */
  pr?: SpacingWithoutInsetProp;
  /** Padding-bottom from theme */
  pb?: SpacingWithoutInsetProp;
  /** Padding-left from theme */
  pl?: SpacingWithoutInsetProp;
  /** Padding on the x-axis (left & right) from theme */
  px?: SpacingWithoutInsetProp;
  /** Padding on the y-axis (top & bottom) from theme */
  py?: SpacingWithoutInsetProp;
}
export interface MarginProps {
  /** Margin from theme */
  margin?: SpacingWithoutInsetProp;
  /** Margin-top from theme */
  mt?: SpacingWithoutInsetProp;
  /** Margin-right from theme */
  mr?: SpacingWithoutInsetProp;
  /** Margin-bottom from theme */
  mb?: SpacingWithoutInsetProp;
  /** Margin-left from theme */
  ml?: SpacingWithoutInsetProp;
  /** Margin on the x-axis (left & right) from theme */
  mx?: SpacingWithoutInsetProp;
  /** Margin on the y-axis (top & bottom) from theme */
  my?: SpacingWithoutInsetProp;
}

export interface SpacingProps extends PaddingProps, MarginProps {}

export const spacingStyles = (props: SpacingProps | undefined) => {
  if (!props) return;
  const array = Object.entries(props).map(([key, value]) =>
    getSpacingStyle(defaultSuomifiTheme, key as keyof SpacingProps, value),
  );
  return Object.assign({}, ...array);
};

export const buildSpacingCSS = (props: SpacingProps | undefined): string => {
  if (!props) return '';

  const cssStyles = Object.entries(props)
    .map(([key, value]) =>
      getCSSSpacing(defaultSuomifiTheme, key as keyof SpacingProps, value),
    )
    .join('');

  return cssStyles;
};

const inlineStyle = {
  margin: 'margin',
  mt: 'marginTop',
  mr: 'marginRight',
  mb: 'marginBottom',
  ml: 'marginLeft',
  padding: 'padding',
  pt: 'paddingTop',
  pr: 'paddingRight',
  pb: 'paddingBottom',
  pl: 'paddingLeft',
};

const cssSelector = {
  margin: 'margin',
  mt: 'margin-top',
  mr: 'margin-right',
  mb: 'margin-bottom',
  ml: 'margin-left',
  padding: 'padding',
  pt: 'padding-top',
  pr: 'padding-right',
  pb: 'padding-bottom',
  pl: 'padding-left',
};

const getSpacingStyle = (
  theme: SuomifiTheme,
  key: keyof SpacingProps,
  value: SpacingProp,
) => {
  const amount = spaceVal(theme)(value);
  switch (key) {
    case 'mx':
      return { marginRight: `${amount}`, marginLeft: `${amount}` };
    case 'my':
      return { marginTop: `${amount}`, marginBottom: `${amount}` };
    case 'px':
      return { paddingRight: `${amount}`, paddingLeft: `${amount}` };
    case 'py':
      return { paddingTop: `${amount}`, paddingBottom: `${amount}` };
    case 'margin':
    case 'mt':
    case 'mr':
    case 'mb':
    case 'ml':
    case 'padding':
    case 'pt':
    case 'pr':
    case 'pb':
    case 'pl':
      return { [inlineStyle[key]]: `${amount}` };
    default:
      return '';
  }
};

const getCSSSpacing = (
  theme: SuomifiTheme,
  key: keyof SpacingProps,
  value: SpacingProp,
) => {
  const amount = spaceVal(theme)(value);
  switch (key) {
    case 'mx':
      return `margin-right: ${amount}; margin-left: ${amount};`;
    case 'my':
      return `margin-top: ${amount}; margin-bottom: ${amount};`;
    case 'px':
      return `padding-right: ${amount}; padding-left: ${amount};`;
    case 'py':
      return `padding-top: ${amount}; padding-bottom: ${amount};`;
    default:
      return `${[cssSelector[key]]}: ${amount};`;
  }
};

export const separateMarginProps = <T extends MarginProps>(
  props: T,
): [MarginProps, Omit<T, keyof MarginProps>] => {
  const { margin, my, mx, mt, mr, mb, ml, ...otherProps } = props;
  return [
    {
      ...(margin !== undefined && { margin }),
      ...(my !== undefined && { my }),
      ...(mx !== undefined && { mx }),
      ...(mt !== undefined && { mt }),
      ...(mr !== undefined && { mr }),
      ...(mb !== undefined && { mb }),
      ...(ml !== undefined && { ml }),
    },
    otherProps,
  ];
};

export const separateMarginAndPaddingProps = <T extends SpacingProps>(
  props: T,
): [SpacingProps, Omit<T, keyof SpacingProps>] => {
  const {
    margin,
    padding,
    my,
    mx,
    py,
    px,
    mt,
    mr,
    mb,
    ml,
    pt,
    pr,
    pb,
    pl,
    ...otherProps
  } = props;
  return [
    {
      ...(margin !== undefined && { margin }),
      ...(padding !== undefined && { padding }),
      ...(my !== undefined && { my }),
      ...(mx !== undefined && { mx }),
      ...(py !== undefined && { py }),
      ...(px !== undefined && { px }),
      ...(mt !== undefined && { mt }),
      ...(mr !== undefined && { mr }),
      ...(mb !== undefined && { mb }),
      ...(ml !== undefined && { ml }),
      ...(pt !== undefined && { pt }),
      ...(pr !== undefined && { pr }),
      ...(pb !== undefined && { pb }),
      ...(pl !== undefined && { pl }),
    },
    otherProps,
  ];
};
