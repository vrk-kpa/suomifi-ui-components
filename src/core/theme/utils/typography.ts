import {
  cssObjectsToCss,
  FlattenSimpleInterpolation,
  cssValueToString,
} from '../../../utils/css';
import { tokens, TypograhpyDesingTokens } from 'suomifi-design-tokens';
import { SuomifiTheme } from '../';

export type TypographyTokensAsCss = {
  [key in keyof typeof tokens.typography]: string
};

type TypographyTokensAsCssProp = {
  [key in keyof TypograhpyDesingTokens]: FlattenSimpleInterpolation
};
export interface TypographyUtil extends TypographyTokensAsCssProp {}
export class TypographyUtil {
  static instance: TypographyUtil;
  constructor(tokens: TypograhpyDesingTokens) {
    // If instance not created, does not take on account if tokens is different!
    if (!TypographyUtil.instance) {
      // Assing typographyTokens as CSS FlattenSimpleInterpolation to this object
      Object.assign(this, cssObjectsToCss(tokens));
      TypographyUtil.instance = this;
    }
    return TypographyUtil.instance;
  }
}

/**
 * Add font-declarations to typography tokens
 * @param typography Typography-tokens
 * @return typographyWithUtils Typography-tokens and typography-tokens as CSS strings
 */
export const typographyUtils = (typography: TypograhpyDesingTokens) => {
  const instance = new TypographyUtil(typography);
  Object.freeze(instance);
  return instance;
};

type ValueTypographyTokenProp = keyof typeof tokens.typography;
export const fontSize = (props: {
  theme: SuomifiTheme;
  [key: string]: any | any[];
}) => (typographyToken: ValueTypographyTokenProp) => {
  return cssValueToString(
    props.theme.values.typography[typographyToken].fontSize,
  );
};
