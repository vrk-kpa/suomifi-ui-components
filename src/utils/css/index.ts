export { alphaHex, hslaToHex } from './colors';
export { allStates, absolute } from './pseudo';
export {
  clearfix,
  cssObjectToCss,
  cssObjectsToCss,
  FlattenSimpleInterpolation,
  cssValueToString,
} from './utils';
