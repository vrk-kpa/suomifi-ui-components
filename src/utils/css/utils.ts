import { css, FlattenSimpleInterpolation } from 'styled-components';
import { ValueUnit } from 'suomifi-design-tokens';
export { FlattenSimpleInterpolation };

export const clearfix = css`
  &:after {
    display: block;
    content: '';
    clear: both;
  }
`;

/**
 * Return FlattenSimpleInterpolation compatible string or number
 * @param cssValue compatible value or {value, unit} pair-object
 */
const cssValueToString = (cssValue: number | string | ValueAndUnit) => {
  if (!!cssValue && typeof cssValue === 'object' && 'value' in cssValue) {
    const { value, unit } = cssValue;
    return !!unit ? `${value}${unit}` : value;
  }
  return cssValue;
};

const camelToSnake = (string: string) =>
  string.replace(/[\w]([A-Z])/g, m => `${m[0]}-${m[1]}`).toLowerCase();
/**
 * Convert CSSObject to CSS FlattenSimpleInterpolation
 * @param value CSSObject
 */
export const cssObjectToCss = <T>(value: T) => css`
  ${Object.entries(value)
    .map(([key, value]) => `${camelToSnake(key)}: ${cssValueToString(value)};`)
    .join('')}
`;

/**
 * Convert CSSObject tokens to CSS FlattenSimpleInterpolation
 * @param tokens Tokens of CSSObjects
 */
export const cssObjectsToCss = <T, K extends keyof T>(tokens: T) =>
  Object.entries(tokens).reduce(
    (retObj, [key, value]) => ({
      ...retObj,
      [key]: cssObjectToCss(value),
    }),
    {} as { [key in K]: FlattenSimpleInterpolation },
  );
