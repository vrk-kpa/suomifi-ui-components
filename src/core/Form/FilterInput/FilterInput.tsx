import React, {
  Component,
  ChangeEvent,
  forwardRef,
  ReactNode,
  ReactElement,
} from 'react';
import { default as styled } from 'styled-components';
import classnames from 'classnames';
import { InputStatus, StatusTextCommonProps } from '../types';
import {
  HtmlInputProps,
  HtmlDiv,
  HtmlDivProps,
  HtmlInput,
} from '../../../reset';
import { AutoId } from '../../utils/AutoId/AutoId';
import { SuomifiThemeProp, SuomifiThemeConsumer } from '../../theme';
import { getConditionalAriaProp } from '../../../utils/aria';
import { Label, LabelMode } from '../Label/Label';
import { HintText } from '../HintText/HintText';
import { StatusText } from '../StatusText/StatusText';
import { baseStyles } from './FilterInput.baseStyles';

const baseClassName = 'fi-filter-input';
const filterInputClassNames = {
  error: `${baseClassName}--error`,
  disabled: `${baseClassName}--disabled`,
  labelAlignLeft: `${baseClassName}--label-align-left`,
  labelIsVisible: `${baseClassName}_label--visible`,
  wrapper: `${baseClassName}_wrapper`,
  functionalityContainer: `${baseClassName}_functionalityContainer`,
  inputElementContainer: `${baseClassName}_input-element-container`,
  inputElement: `${baseClassName}_input`,
  actionElementsContainer: `${baseClassName}_action-elements-container`,
  statusTextHasContent: `${baseClassName}_statusText--has-content`,
};

export type FilterInputStatus = Exclude<InputStatus, 'success'>;

interface InternalFilterInputProps<T>
  extends Omit<HtmlInputProps, 'type' | 'onChange'>,
    StatusTextCommonProps {
  /** FilterInput container div class name for custom styling. */
  className?: string;
  /** FilterInput container div props */
  inputContainerProps?: Omit<HtmlDivProps, 'className'>;
  /** FilterInput element container div props, e.g. for widget roles */
  inputElementContainerProps?: Omit<HtmlDivProps, 'className'>;
  /** Disable input usage */
  disabled?: boolean;
  /** Placeholder text for input. Use only as visual aid, not for instructions. */
  visualPlaceholder?: string;
  /** Label */
  labelText: ReactNode;
  /** Hide or show label. Label element is always present, but can be visually hidden.
   * @default visible
   */
  labelMode?: LabelMode;
  /** Text to mark a field optional. Will be wrapped in parentheses and shown after labelText. */
  optionalText?: string;
  /** Hint text to be shown below the label */
  hintText?: string;
  /**
   * 'default' | 'error'
   * @default default
   */
  status?: FilterInputStatus;
  /** FilterInput name */
  name?: string;
  /** Align label on top or on the left side of the input field
   * @default 'top'
   */
  labelAlign?: 'top' | 'left';
  /** Items to be filtered */
  items: Array<T>;
  /** Callback for items filtering event  */
  onFilter: (filteredItems: Array<T>) => void;
  /** Filtering rule to be used */
  filterFunc: (item: T, query: string) => boolean;
  /** Input value onChange handler */
  onChange?: (value: string) => void;
  /** Children for the input container element. Renders inside and on top of the input. Aligns to right. */
  children?: ReactNode;
  /** Tooltip component for the input's label */
  tooltipComponent?: ReactElement;
  shouldFilter?: boolean;
}

interface InnerRef {
  forwardedRef: React.RefObject<HTMLInputElement>;
}

export interface FilterInputProps extends InternalFilterInputProps<any> {
  /** Ref object for the input element */
  ref?: React.RefObject<HTMLInputElement>;
}

class BaseFilterInput<T> extends Component<FilterInputProps & InnerRef> {
  componentDidUpdate(prevProps: FilterInputProps) {
    if (
      (!!this.props.onFilter &&
        prevProps.value !== this.props.value &&
        this.props.shouldFilter) ||
      prevProps.items !== this.props.items
    ) {
      const value = !!this.props.value ? this.props.value.toString() : '';
      this.props.onFilter(
        this.filterItems(this.props.items, this.props.filterFunc, value),
      );
    }
  }

  private filterItems = (
    items: Array<T>,
    filterFunc: (item: T, query: string) => boolean,
    value: string,
  ) =>
    items.reduce((filtered: T[], item: T) => {
      if (filterFunc(item, value)) {
        filtered.push(item);
      }
      return filtered;
    }, []);

  render() {
    const {
      className,
      inputContainerProps,
      inputElementContainerProps,
      visualPlaceholder,
      labelText,
      labelMode,
      hintText,
      optionalText,
      status,
      statusText,
      id,
      labelAlign,
      'aria-describedby': ariaDescribedBy,
      statusTextAriaLiveMode = 'assertive',
      items: propItems,
      onFilter: propOnFiltering,
      filterFunc: propFilterRule,
      forwardedRef,
      onChange: propOnChange,
      children,
      tooltipComponent,
      shouldFilter,
      ...passProps
    } = this.props;

    const onChangeHandler = (event: ChangeEvent<HTMLInputElement>) => {
      const { items, onFilter, filterFunc } = this.props;
      const { value: eventValue } = event.target;

      const filteredItems: T[] = this.filterItems(
        items,
        filterFunc,
        eventValue,
      );

      onFilter(filteredItems);
      if (propOnChange) {
        propOnChange(eventValue);
      }
    };

    const labelId = `${id}-label`;
    const hintTextId = hintText ? `${id}-hintText` : undefined;
    const statusTextId = statusText ? `${id}-statusText` : undefined;

    return (
      <HtmlDiv
        {...inputContainerProps}
        className={classnames(baseClassName, className, {
          [filterInputClassNames.disabled]: !!passProps.disabled,
          [filterInputClassNames.error]: status === 'error',
          [filterInputClassNames.labelAlignLeft]: labelAlign === 'left',
        })}
      >
        <HtmlDiv className={classnames(filterInputClassNames.wrapper, {})}>
          <Label
            id={labelId}
            labelMode={labelMode}
            optionalText={optionalText}
            className={classnames({
              [filterInputClassNames.labelIsVisible]: labelMode !== 'hidden',
            })}
            tooltipComponent={tooltipComponent}
            htmlFor={id}
          >
            {labelText}
          </Label>
          <HintText id={hintTextId}>{hintText}</HintText>
          <HtmlDiv className={filterInputClassNames.functionalityContainer}>
            <HtmlDiv
              className={filterInputClassNames.inputElementContainer}
              {...inputElementContainerProps}
            >
              <HtmlInput
                {...passProps}
                aria-labelledby={labelId}
                aria-invalid={status === 'error'}
                id={id}
                className={filterInputClassNames.inputElement}
                type="text"
                forwardedRef={forwardedRef}
                placeholder={visualPlaceholder}
                {...getConditionalAriaProp('aria-describedby', [
                  statusTextId,
                  hintTextId,
                  ariaDescribedBy,
                ])}
                autoComplete="off"
                aria-autocomplete="list"
                autoCapitalize="none"
                spellCheck="false"
                onChange={onChangeHandler}
                aria-multiline={false}
              />
            </HtmlDiv>
            {React.Children.count(children) > 0 && (
              <HtmlDiv
                className={filterInputClassNames.actionElementsContainer}
                onMouseDownCapture={(event) => {
                  event.preventDefault();
                  event.stopPropagation();
                }}
              >
                {children}
              </HtmlDiv>
            )}
            <StatusText
              id={statusTextId}
              className={classnames({
                [filterInputClassNames.statusTextHasContent]: !!statusText,
              })}
              status={status}
              disabled={passProps.disabled}
              ariaLiveMode={statusTextAriaLiveMode}
            >
              {statusText}
            </StatusText>
          </HtmlDiv>
        </HtmlDiv>
      </HtmlDiv>
    );
  }
}
/** This wrapper is needed to make TypeScript work with styled components and generics */
const BaseFilterInputWrapper: <T>(
  props: InternalFilterInputProps<T> & InnerRef & SuomifiThemeProp,
) => JSX.Element = ({
  // eslint-disable-next-line react/prop-types
  theme,
  ...passProps
}) => <BaseFilterInput {...passProps} />;

const StyledFilterInput = styled(BaseFilterInputWrapper)`
  ${({ theme }) => baseStyles(theme)}
`;

/**
 * <i class="semantics" />
 * Use for filtering.
 * Props other than specified explicitly are passed on to underlying input element.
 */
const FilterInput = forwardRef(
  (props: FilterInputProps, ref: React.RefObject<HTMLInputElement>) => {
    const { id: propId, ...passProps } = props;
    return (
      <SuomifiThemeConsumer>
        {({ suomifiTheme }) => (
          <AutoId id={propId}>
            {(id) => (
              <StyledFilterInput
                theme={suomifiTheme}
                id={id}
                forwardedRef={ref}
                {...passProps}
              />
            )}
          </AutoId>
        )}
      </SuomifiThemeConsumer>
    );
  },
);

FilterInput.displayName = 'FilterInput';
export { FilterInput };
