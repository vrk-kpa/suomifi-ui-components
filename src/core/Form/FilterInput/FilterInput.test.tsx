import React from 'react';
import { render, fireEvent } from '@testing-library/react';
import { axeTest } from '../../../utils/test';
import { FilterInput } from './FilterInput';

interface Product {
  name: string;
  price: number;
  tax: boolean;
}

const TestFilterInput = (
  <FilterInput
    labelText="Label"
    items={[]}
    onFilter={jest.fn()}
    filterFunc={() => true}
  />
);

test('should not have basic accessibility issues', async () =>
  axeTest(TestFilterInput));

test('snapshot matches', () => {
  const { container } = render(TestFilterInput);
  expect(container.firstChild).toMatchSnapshot();
});

describe('props', () => {
  const mockOnFilter = jest.fn();
  const mockFilterFunc = jest.fn(() => true);

  describe('className', () => {
    it('has the given custom className', () => {
      const { container } = render(
        <FilterInput
          className="custom-style"
          labelText="Label"
          items={[]}
          onFilter={mockOnFilter}
          filterFunc={mockFilterFunc}
        />,
      );
      expect(container.firstChild).toHaveClass('custom-style');
    });
  });

  describe('status', () => {
    it('has the className when in error state', () => {
      const { container } = render(
        <FilterInput
          className="custom-style"
          labelText="Label"
          status="error"
          items={[]}
          onFilter={mockOnFilter}
          filterFunc={mockFilterFunc}
        />,
      );
      expect(container.firstChild).toHaveClass('fi-filter-input--error');
    });
  });

  describe('statusText', () => {
    it('has the status text element', () => {
      const { getByText } = render(
        <FilterInput
          statusText="Example status text"
          labelText="Label"
          items={[]}
          onFilter={mockOnFilter}
          filterFunc={mockFilterFunc}
        />,
      );
      const statusText = getByText('Example status text');
      expect(statusText).toHaveClass('fi-status-text');
    });

    it('will be added to input aria-describedby', () => {
      const { getByRole } = render(
        <FilterInput
          statusText="Example status text"
          id="123"
          labelText="Label"
          items={[]}
          onFilter={mockOnFilter}
          filterFunc={mockFilterFunc}
        />,
      );
      expect(getByRole('textbox')).toHaveAttribute(
        'aria-describedby',
        '123-statusText',
      );
    });
  });

  describe('disabled', () => {
    it('has disabled attribute and classname', () => {
      const { container, getByRole } = render(
        <FilterInput
          disabled={true}
          labelText="Label"
          items={[]}
          onFilter={mockOnFilter}
          filterFunc={mockFilterFunc}
        />,
      );
      expect(container.firstChild).toHaveClass('fi-filter-input--disabled');

      const inputField = getByRole('textbox') as HTMLInputElement;
      expect(inputField).toHaveAttribute('disabled');
    });
  });

  describe('labelAlign', () => {
    it('has the className for left align', () => {
      const { container } = render(
        <FilterInput
          labelAlign="left"
          labelText="Label"
          items={[]}
          onFilter={mockOnFilter}
          filterFunc={mockFilterFunc}
        />,
      );
      expect(container.firstChild).toHaveClass(
        'fi-filter-input--label-align-left',
      );
    });
  });

  describe('labelText', () => {
    it('should be found ', () => {
      const { getByText } = render(
        <FilterInput
          labelText="Label"
          items={[]}
          onFilter={mockOnFilter}
          filterFunc={mockFilterFunc}
        />,
      );
      const label = getByText('Label');
      expect(label).toHaveClass('fi-label_label-span');
    });
  });

  describe('optionalText', () => {
    it('should have element and correct classname for it', () => {
      const { getByText } = render(
        <FilterInput
          labelText="Label"
          optionalText="Optional"
          items={[]}
          onFilter={mockOnFilter}
          filterFunc={mockFilterFunc}
        />,
      );
      const optionalText = getByText('(Optional)');
      expect(optionalText).toHaveClass('fi-label_optional-text');
    });
  });

  describe('labelMode', () => {
    it('should be visible by default', () => {
      const { getByText } = render(
        <FilterInput
          labelText="Label"
          optionalText="Optional"
          items={[]}
          onFilter={mockOnFilter}
          filterFunc={mockFilterFunc}
        />,
      );
      const label = getByText('Label');
      expect(label).toHaveClass('fi-label_label-span');
    });

    it('should be hidden', () => {
      const { getByText } = render(
        <FilterInput
          labelText="Label"
          labelMode="hidden"
          items={[]}
          onFilter={mockOnFilter}
          filterFunc={mockFilterFunc}
        />,
      );
      const label = getByText('Label');
      expect(label).toHaveClass('fi-visually-hidden');
    });
  });

  describe('visualPlaceholder', () => {
    it('should have the given text', () => {
      const { getByRole } = render(
        <FilterInput
          labelText="Label"
          visualPlaceholder="Enter text here"
          items={[]}
          onFilter={mockOnFilter}
          filterFunc={mockFilterFunc}
        />,
      );
      const inputField = getByRole('textbox') as HTMLInputElement;
      expect(inputField).toHaveAttribute('placeholder', 'Enter text here');
    });
  });

  describe('onFilter', () => {
    const tools: Product[] = [
      { name: 'Jackhammer', price: 230, tax: false },
      { name: 'Hammer', price: 15, tax: true },
      { name: 'Sledgehammer', price: 36, tax: false },
      { name: 'Spade', price: 50, tax: true },
      { name: 'Powersaw', price: 150, tax: false },
    ];

    it('should return filtered items', () => {
      const mockedOnFilter = jest.fn((filtered) => filtered);
      const filter = (tool: Product, query: string) =>
        tool.name.includes(query);
      const { getByRole } = render(
        <FilterInput
          labelText="Label"
          visualPlaceholder="Enter text here"
          items={tools}
          onFilter={mockedOnFilter}
          filterFunc={filter}
          shouldFilter={true}
        />,
      );
      const inputField = getByRole('textbox') as HTMLInputElement;
      fireEvent.change(inputField, { target: { value: 'hammer' } });
      expect(mockedOnFilter.mock.results).toEqual([
        {
          type: 'return',
          value: [
            { name: 'Jackhammer', price: 230, tax: false },
            { name: 'Sledgehammer', price: 36, tax: false },
          ],
        },
      ]);
    });
  });
});
