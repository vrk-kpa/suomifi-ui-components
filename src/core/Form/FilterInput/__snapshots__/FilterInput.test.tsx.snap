// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot matches 1`] = `
.c0 {
  line-height: 1.15;
  -ms-text-size-adjust: 100%;
  -webkit-text-size-adjust: 100%;
  margin: 0;
  padding: 0;
  border: 0;
  box-sizing: border-box;
  font: 100% inherit;
  line-height: 1;
  text-align: left;
  text-decoration: none;
  vertical-align: baseline;
  color: inherit;
  background: none;
  cursor: inherit;
  display: block;
  max-width: 100%;
  word-wrap: normal;
  word-break: normal;
  white-space: normal;
}

.c0 ::before,
.c0 ::after {
  box-sizing: border-box;
}

.c2 {
  line-height: 1.15;
  -ms-text-size-adjust: 100%;
  -webkit-text-size-adjust: 100%;
  margin: 0;
  padding: 0;
  border: 0;
  box-sizing: border-box;
  font: 100% inherit;
  line-height: 1;
  text-align: left;
  text-decoration: none;
  vertical-align: baseline;
  color: inherit;
  background: none;
  cursor: inherit;
  display: block;
  max-width: 100%;
  word-wrap: normal;
  word-break: normal;
  white-space: normal;
}

.c2 ::before,
.c2 ::after {
  box-sizing: border-box;
}

.c5 {
  line-height: 1.15;
  -ms-text-size-adjust: 100%;
  -webkit-text-size-adjust: 100%;
  margin: 0;
  overflow: visible;
  margin: 0;
  padding: 0;
  border: 0;
  box-sizing: border-box;
  font: 100% inherit;
  line-height: 1;
  text-align: left;
  text-decoration: none;
  vertical-align: baseline;
  color: inherit;
  background: none;
  cursor: inherit;
  display: inline-block;
  max-width: 100%;
}

.c5::-webkit-input-placeholder {
  color: inherit;
  opacity: 0.54;
}

.c5 ::before,
.c5 ::after {
  box-sizing: border-box;
}

.c4 {
  line-height: 1.15;
  -ms-text-size-adjust: 100%;
  -webkit-text-size-adjust: 100%;
  margin: 0;
  padding: 0;
  border: 0;
  box-sizing: border-box;
  font: 100% inherit;
  line-height: 1;
  text-align: left;
  text-decoration: none;
  vertical-align: baseline;
  color: inherit;
  background: none;
  cursor: inherit;
  display: inline;
  max-width: 100%;
  word-wrap: normal;
  word-break: normal;
  white-space: normal;
}

.c4 ::before,
.c4 ::after {
  box-sizing: border-box;
}

.c3.fi-label .fi-label_label-span {
  letter-spacing: 0;
  text-decoration: none;
  word-break: break-word;
  overflow-wrap: break-word;
  -webkit-font-smoothing: antialiased;
  font-family: 'Source Sans Pro','Helvetica Neue','Arial',sans-serif;
  font-size: 16px;
  line-height: 1.5;
  font-weight: 600;
  display: inline-block;
  vertical-align: middle;
  color: hsl(0, 0%, 13%);
}

.c3.fi-label .fi-label_label-span .fi-label_optional-text {
  font-family: 'Source Sans Pro','Helvetica Neue','Arial',sans-serif;
  font-size: 16px;
  line-height: 1.5;
  font-weight: 400;
}

.c3.fi-label .fi-label_label-span .fi-tooltip {
  margin-left: 6px;
}

.c6 {
  letter-spacing: 0;
  text-decoration: none;
  word-break: break-word;
  overflow-wrap: break-word;
  -webkit-font-smoothing: antialiased;
  font-family: 'Source Sans Pro','Helvetica Neue','Arial',sans-serif;
  font-size: 16px;
  line-height: 1.5;
  font-weight: 600;
  color: hsl(0, 0%, 13%);
  font-size: 14px;
  line-height: 20px;
}

.c6.fi-status-text {
  display: block;
}

.c6.fi-status-text.fi-status-text--error {
  color: hsl(3, 59%, 43%);
}

.c6.fi-status-text .fi-icon {
  vertical-align: middle;
  transform: translateY(-0.1em);
  margin-right: 5px;
}

.c6.fi-status-text.fi-status-text--success .fi-icon-base-fill {
  fill: hsl(166, 90%, 34%);
}

.c1 {
  letter-spacing: 0;
  text-decoration: none;
  word-break: break-word;
  overflow-wrap: break-word;
  -webkit-font-smoothing: antialiased;
  font-family: 'Source Sans Pro','Helvetica Neue','Arial',sans-serif;
  font-size: 18px;
  line-height: 1.5;
  font-weight: 400;
}

.c1 .fi-filter-input_wrapper {
  display: inline-block;
  width: 100%;
}

.c1 .fi-filter-input_wrapper .fi-filter-input_label--visible {
  margin-bottom: 10px;
}

.c1 .fi-filter-input_wrapper .fi-hint-text {
  margin-bottom: 10px;
}

.c1 .fi-filter-input_functionalityContainer {
  line-height: 1.5;
  position: relative;
}

.c1 .fi-filter-input_functionalityContainer .fi-filter-input_action-elements-container {
  position: absolute;
  height: 40px;
  right: 0;
  top: 0;
  display: flex;
  flex-direction: row;
  align-items: center;
}

.c1 .fi-filter-input_functionalityContainer .fi-filter-input_action-elements-container>* {
  box-sizing: content-box;
  display: flex;
  padding: 0 8px;
  flex: 0 0 auto;
}

.c1 .fi-filter-input_functionalityContainer .fi-filter-input_action-elements-container :not(:last-child) {
  border-right: 1px solid hsl(202, 7%, 67%);
  padding-right: 8px;
  height: 20px;
}

.c1 .fi-filter-input_functionalityContainer .fi-filter-input_statusText--has-content {
  margin-top: 5px;
}

.c1 .fi-filter-input_input-element-container >input:focus {
  outline-color: hsl(196, 77%, 44%);
  outline-width: 2px;
  outline-offset: 2px;
  outline-style: solid;
}

.c1 .fi-filter-input_input-element-container:focus-within >input:focus {
  outline: none;
}

.c1 .fi-filter-input_input-element-container:focus-within {
  position: relative;
}

.c1 .fi-filter-input_input-element-container:focus-within::after {
  content: '';
  position: absolute;
  pointer-events: none;
  top: -2px;
  right: -2px;
  bottom: -2px;
  left: -2px;
  border-radius: 2px;
  background-color: transparent;
  border: 0px solid hsl(0, 0%, 100%);
  box-sizing: border-box;
  box-shadow: 0 0 0 2px hsl(196, 77%, 44%);
  z-index: 9999;
}

.c1 .fi-filter-input_input {
  color: hsl(0, 0%, 13%);
  letter-spacing: 0;
  text-decoration: none;
  word-break: break-word;
  overflow-wrap: break-word;
  -webkit-font-smoothing: antialiased;
  font-family: 'Source Sans Pro','Helvetica Neue','Arial',sans-serif;
  font-size: 16px;
  line-height: 1.5;
  font-weight: 400;
  min-width: 245px;
  max-width: 100%;
  padding: 8px 16px;
  border: 1px solid hsl(202, 7%, 80%);
  border-radius: 2px;
  line-height: 1;
  background-color: hsl(0, 0%, 100%);
  min-width: 40px;
  width: 100%;
  min-height: 40px;
  padding-left: 10px;
  border-color: hsl(201, 7%, 58%);
}

.c1 .fi-filter-input_input::placeholder {
  font-style: italic;
  color: hsl(201, 7%, 46%);
  opacity: 1;
}

.c1.fi-filter-input--error .fi-filter-input_input {
  border-color: hsl(3, 59%, 43%);
  border-width: 2px;
}

.c1.fi-filter-input--disabled .fi-filter-input_input {
  color: hsl(202, 7%, 67%);
  background-color: hsl(202, 7%, 97%);
  border-color: hsl(202, 7%, 80%);
  cursor: not-allowed;
}

.c1.fi-filter-input--label-align-left .fi-filter-input_wrapper {
  display: flex;
  flex-direction: row;
  align-items: flex-start;
}

.c1.fi-filter-input--label-align-left .fi-filter-input_wrapper .fi-label {
  padding-right: 10px;
  padding-top: 8px;
  align-self: flex-start;
}

.c1.fi-filter-input--label-align-left .fi-filter-input_wrapper .fi-label .fi-label_label-span {
  margin-bottom: 0;
}

<div
  class="c0 fi-filter-input c1"
>
  <div
    class="c0 fi-filter-input_wrapper"
  >
    <div
      class="c2 c3 fi-filter-input_label--visible fi-label"
    >
      <label
        class="c4 fi-label_label-span"
        for="1"
        id="1-label"
      >
        Label
      </label>
    </div>
    <div
      class="c0 fi-filter-input_functionalityContainer"
    >
      <div
        class="c0 fi-filter-input_input-element-container"
      >
        <input
          aria-autocomplete="list"
          aria-invalid="false"
          aria-labelledby="1-label"
          aria-multiline="false"
          autocapitalize="none"
          autocomplete="off"
          class="c5 fi-filter-input_input"
          id="1"
          spellcheck="false"
          type="text"
        />
      </div>
      <span
        aria-atomic="true"
        aria-live="assertive"
        class="c4 c6 fi-status-text"
      />
    </div>
  </div>
</div>
`;
